compiler/noirc_evaluator/src/ssa/opt/inline_const_brillig_calls.rs

```diff
@@ -0,0 +1,220 @@
+        let TerminatorInstruction::Return { return_values, call_stack: _ } = terminator else {
+            brillig_functions_we_could_not_inline.insert(*func_id);
+            return false;
+        };
+
+        // Sanity check: make sure all returned values are constant
+        if !return_values.iter().all(|value_id| function.dfg.is_constant(*value_id)) {
+            brillig_functions_we_could_not_inline.insert(*func_id);
+            return false;
+        }
+
+        // Replace the instruction results with the constant values we got
+        let current_results = self.dfg.instruction_results(instruction_id).to_vec();
+        assert_eq!(return_values.len(), current_results.len());
+
+        for (current_result_id, return_value_id) in current_results.iter().zip(return_values) {
+            let new_return_value_id = copy_constant_to_function(&function, self, return_value_id);
+            self.dfg.set_value_from_id(*current_result_id, new_return_value_id);
+        }
+
+        true
+    }
+}
+
+/// Copies a constant from one function to another.
+/// Though it might seem we can just take a value out of `from_function` and call `make_value` on `to_function`,
+/// if the constant is an array the values will still keep pointing to `from_function`. So, this function
+/// recursively copies the array values too.
+fn copy_constant_to_function(
+    from_function: &Function,
+    to_function: &mut Function,
+    constant_id: ValueId,
+) -> ValueId {
+    if let Some((constant, typ)) = from_function.dfg.get_numeric_constant_with_type(constant_id) {
+        to_function.dfg.make_constant(constant, typ)
+    } else if let Some((constants, typ)) = from_function.dfg.get_array_constant(constant_id) {
+        let new_constants = constants
+            .iter()
+            .map(|constant_id| copy_constant_to_function(from_function, to_function, *constant_id))
+            .collect();
+        to_function.dfg.make_array(new_constants, typ)
+    } else {
+        unreachable!("A constant should be either a numeric constant or an array constant")
+    }
+}
+
+/// Optimizes a function by running the same passes as `optimize_into_acir`
+/// after the `inline_const_brillig_calls` pass.
+/// The function is changed to be an ACIR function so the function can potentially
+/// be optimized into a single return terminator.
+fn optimize(
+    mut function: Function,
+    inliner_aggressiveness: i64,
+    error_selector_to_type: &BTreeMap<ErrorSelector, Type>,
+) -> Result<Function, RuntimeError> {
+    function.set_runtime(RuntimeType::Acir(InlineType::InlineAlways));
+
+    let ssa = Ssa::new(vec![function], error_selector_to_type.clone());
+    let builder = SsaBuilder { ssa, print_ssa_passes: false, print_codegen_timings: false };
+    let mut ssa = optimize_ssa_after_inline_const_brillig_calls(
+        builder,
+        inliner_aggressiveness,
+        false, // don't inline functions with no predicates
```

Comment:
It would be nice to have a comment on why not? Asking for a friend ;)

Reply:
😆 

The main reason is that running that pass, it crashed. But then I found that it crashed because to inline things we need to have all other functions in context, and here we just have this single function in context. I'll add a comment for this.